pkg/R/DataFrame.R

```diff
@@ -114,20 +114,192 @@ setMethod("count",
+# but allow for use with DataFrames by first converting to an RRDD before calling #
+# the requested map function.                                                     #
+###################################################################################
+
+setMethod("lapply",
+          signature(X = "DataFrame", FUN = "function"),
+          function(X, FUN) {
+            rdd <- toRDD(X)
+            lapply(rdd, FUN)
+          })
+
+setMethod("map",
+          signature(X = "DataFrame", FUN = "function"),
+          function(X, FUN) {
+            lapply(X, FUN)
+          })
+
+setMethod("flatMap",
+          signature(X = "DataFrame", FUN = "function"),
+          function(X, FUN) {
+            rdd <- toRDD(X)
+            flatMap(rdd, FUN)
+          })
+
+setMethod("lapplyPartition",
+          signature(X = "DataFrame", FUN = "function"),
+          function(X, FUN) {
+            rdd <- toRDD(X)
+            lapplyPartition(rdd, FUN)
+          })
+
+setMethod("mapPartitions",
+          signature(X = "DataFrame", FUN = "function"),
+          function(X, FUN) {
+            lapplyPartition(X, FUN)
+          })
+
+setMethod("lapplyPartitionsWithIndex",
+          signature(X = "DataFrame", FUN = "function"),
+          function(X, FUN) {
+            rdd <- toRDD(X)
+            lapplyPartitionsWithIndex(rdd, FUN)
+          })
+
+setMethod("mapPartitionsWithIndex",
+          signature(X = "DataFrame", FUN = "function"),
+          function(X, FUN){
+            lapplyPartitionsWithIndex(X, FUN)
+          })
+
+setMethod("foreach",
+          signature(rdd = "DataFrame", func = "function"),
+          function(rdd, func) {
+            rddIn <- toRDD(rdd)
+            foreach(rddIn, func)
+          })
+
+setMethod("foreachPartition",
+          signature(rdd = "DataFrame", func = "function"),
+          function(rdd, func) {
+            rddIn <- toRDD(rdd)
+            foreachPartition(rddIn, func)
+          })
+
+#' Collect to DataFrame
+#' 
+#' Collects all the elements of a Spark DataFrame and coerces them into an R data.frame.
+#' 
+#' @param df A Spark DataFrame
+#' @return An R data.frame
+#' 
+#' @rdname collectToDF
+#' @export
+#' @examples
+#'\dontrun{
+#' sc <- sparkR.init()
+#' sqlCtx <- sparkRSQL.init(sc)
+#' path <- "path/to/file.json"
+#' df <- jsonFile(sqlCtx, path)
+#' converted <- collectToDF(df)
+#' is.data.frame(converted)
+#' }
+
+setGeneric("collectToDF", function(df) { standardGeneric("collectToDF") })
+
+setMethod("collectToDF",
+          signature(df = "DataFrame"),
+          function(df) {
+            listCols <- callJStatic("edu.berkeley.cs.amplab.sparkr.SQLUtils", "dfToColRDD", df@sdf)
+            cols <- lapply(seq_along(listCols),
+                           function(colIdx) {
+                             rddCol <- listCols[[colIdx]]
+                             # Returns a list of byte arrays per partition
+                             rddRaw <- callJMethod(rddCol, "collect")
```

Comment:
@davies Does the new DataFrame API provide any new methods for collecting a DataFrame differently? The problem here is that, since R stores everything as column-major, having to transpose everything when converting to a `data.frame` is really, really inefficient. That said, I know the new API does have a "Column" class. Is there now a built-in way to re-orient the DataFrame so that it's truly stored as columns instead of rows?